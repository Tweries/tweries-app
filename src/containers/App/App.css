@import url('https://fonts.googleapis.com/css?family=Leckerli+One&display=swap');

article,
button {
  color: rgb(20, 23, 26);
}

button,
input,
textarea {
  background-color: rgb(230, 236, 240);
  border: 1px solid rgb(204, 214, 221);
  font-size: 16px;
}

/* --- */

article {
  align-items: center;
  display: flex;
  flex-direction: column;
  justify-content: center;
  padding: 8px;
}

body {
  background-color: rgb(29, 161, 242, 0.125);
}

button:disabled {
  background-color: rgb(204, 214, 221);
}

footer {
  color: red;
  font-size: 12px;
  margin-bottom: 8px;
}

footer.healthy {
  color: green;
}

form {
  display: flex;
  flex-direction: column;
  width: 100%;
}

form > * {
  margin-bottom: 32px;
}

h1 {
  color: rgb(20, 23, 26);
  font-family: 'Leckerli One', cursive;
  font-size: 40px;
  margin: 16px 0;
}

li {
  display: flex;
  flex-direction: column;
  margin-bottom: 32px;
  width: 100%;
}

ul {
  list-style: none;
  margin: 0;
  padding: 0;
}
