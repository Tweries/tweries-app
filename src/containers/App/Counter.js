import React from 'react';

function Counter({ length }) {
  const value = length === 0 ? '' : length;
  return <small className="App__counter">{value}</small>;
}

export default Counter;
