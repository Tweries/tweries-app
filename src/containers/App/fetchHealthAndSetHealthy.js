import fetchHealth from '../../api/fetchHealth';
import { types } from '../../store/makeReducer';

async function fetchHealthAndSetHealthy(dispatch) {
  function setHealthy(error, response) {
    if (error) {
      console.log(error);
      dispatch({ type: types.SET_HEALTHY, value: false });
    }
    if (response) {
      console.log(response);
      dispatch({
        type: types.SET_HEALTHY,
        value: !response.error
      });
    }
  }

  try {
    const response = await fetchHealth();
    setHealthy(null, response);
  } catch (error) {
    setHealthy(error);
  }
}

export default fetchHealthAndSetHealthy;
