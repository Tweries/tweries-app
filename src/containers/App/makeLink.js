function makeLink(data) {
  return `https://twitter.com/${data.userId.split('|')[1]}/status/${
    data.statusIds[0]
  }`;
}

export default makeLink;
