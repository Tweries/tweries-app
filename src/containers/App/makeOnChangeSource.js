import { types } from '../../store/makeReducer';

function makeOnChangeSource({ dispatch, setSource }) {
  function onChangeSource(e) {
    dispatch({
      type: types.CHANGE_SOURCE,
      value: e.target.value
    });
    setSource(e.target.value);
  }

  return onChangeSource;
}

export default makeOnChangeSource;
