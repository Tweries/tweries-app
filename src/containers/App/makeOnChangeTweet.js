import { types } from '../../store/makeReducer';

function makeOnChangeTweet({ dispatch }) {
  function onChangeTweet(e, item) {
    dispatch({
      type: types.CHANGE_TWEET,
      value: {
        id: item.id,
        tweet: e.target.value
      }
    });
  }

  return onChangeTweet;
}

export default makeOnChangeTweet;
