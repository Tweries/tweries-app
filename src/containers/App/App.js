import classnames from 'classnames';
import React, { useEffect, useReducer } from 'react';
import { version } from '../../../package.json';
import fetchHealth from '../../api/fetchHealth.js';
import fetchTweetstorm from '../../api/fetchTweetstorm.js';
import Footer from '../../components/Footer/Footer';
import LinefeedPicker from '../../components/LinefeedPicker/LinefeedPicker';
import NavBar from '../../components/NavBar/NavBar';
import ToastNotification from '../../components/ToastNotification/ToastNotification.js';
import useLocalStorage from '../../hooks/useLocalStorage';
import makeInitialState from '../../store/makeInitialState';
import { types } from '../../store/reducer';
import { NEWLINE } from '../../constants';
import { useAuth0 } from '../../react-auth0-wrapper';
import './App.css';
import Counter from './Counter';

function App({ feature, reducer }) {
  const { isAuthenticated, loading } = useAuth0();

  useEffect(() => {
    async function fetchData() {
      try {
        const data = await fetchHealth();
        console.log(data);
        dispatch({ type: types.SET_HEALTHY, value: true });
      } catch (error) {
        console.log(error);
        dispatch({ type: types.SET_HEALTHY, value: false });
      }
    }
    fetchData();
  }, []);

  const [hashtags_, setHashtags] = useLocalStorage('hashtags', '');
  const [source_, setSource] = useLocalStorage('source', '');

  const [
    { hashtags, healthy, items, notification, source, userId },
    dispatch
  ] = useReducer(
    reducer,
    makeInitialState({
      feature,
      hashtags: hashtags_,
      linefeed: NEWLINE,
      source: source_
    })
  );

  function disabled() {
    return !isAuthenticated || !items.length > 0 || !healthy;
  }

  return loading ? (
    <article>...</article>
  ) : (
    <article>
      <NavBar dispatch={dispatch} />
      <h1>Tweries</h1>
      <form onSubmit={e => e.preventDefault()}>
        <small className="App__small">
          Start typing, to insert a break prior to reaching 280 characters
          please use{' '}
          <LinefeedPicker
            feature={feature}
            onChange={value => {
              dispatch({ type: types.CHANGE_LINEFEED, value });
            }}
          />
        </small>
        <textarea
          data-testid="source"
          placeholder="What's happening?"
          rows={8}
          value={source}
          onChange={e => {
            dispatch({ type: types.CHANGE_SOURCE, value: e.target.value });
            setSource(e.target.value);
          }}
        />
        <Counter length={source.length} />
        <textarea
          data-testid="hashtags"
          onChange={e => {
            dispatch({ type: types.CHANGE_HASHTAGS, value: e.target.value });
            setHashtags(e.target.value);
          }}
          placeholder="#hashtags"
          rows={1}
          type="text"
          value={hashtags}
        />
        <Counter length={hashtags.length} />
        <ul data-testid="list">
          {items.map((item, index) => (
            <li key={index}>
              <textarea readOnly rows={4} value={item.tweet} />
              <Counter length={item.length} />
            </li>
          ))}
        </ul>
        <button
          className={classnames('App__button', {
            'App__button--disabled': disabled()
          })}
          data-testid="tweet"
          disabled={disabled()}
          onClick={async () => {
            try {
              const data = await fetchTweetstorm({ items, userId });
              console.log(data);
              dispatch({
                type: types.RESET_TWEETSTORM,
                value: {
                  message: 'The tweetstorm has been created successfully.',
                  type: 'success'
                }
              });
              setSource('');
              setHashtags('');
            } catch (error) {
              console.log(error);
              dispatch({
                type: types.RESET_TWEETSTORM,
                value: { message: error.message, type: 'danger' }
              });
              setSource('');
              setHashtags('');
            }
          }}
        >
          Tweet
        </button>
      </form>
      <ToastNotification
        notification={notification}
        onClick={() => dispatch({ type: types.DISMISS_TOAST })}
      />
      <Footer healthy={healthy} version={version} />
    </article>
  );
}

export default App;
