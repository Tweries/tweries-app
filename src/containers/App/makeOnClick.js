import fetchTweetstorm from '../../api/fetchTweetstorm';
import { DANGER, SUCCESS } from '../../constants';
import { types } from '../../store/makeReducer';
import makeLink from './makeLink';

function makeOnClick({
  dispatch,
  inReplyToTweetUrl,
  items,
  setInReplyToTweetUrl,
  setSource,
  setWaiting,
  userId
}) {
  function resetTweetstorm(error, response) {
    let link = null;
    let message = null;
    let type = SUCCESS;
    if (error || response.error) {
      message = error ? error.message : response.error.message;
      type = DANGER;
    } else {
      link = makeLink(response.data);
    }
    console.log(error, response);
    dispatch({
      type: types.RESET_TWEETSTORM,
      value: {
        link,
        message,
        type
      }
    });
    if (type === SUCCESS) {
      setInReplyToTweetUrl('');
      setSource('');
    }
    setWaiting(false);
  }

  async function onClick() {
    setWaiting(true);
    try {
      const response = await fetchTweetstorm({
        inReplyToTweetUrl,
        items,
        userId
      });
      resetTweetstorm(null, response);
    } catch (error) {
      resetTweetstorm(error);
    }
  }

  return onClick;
}

export default makeOnClick;
