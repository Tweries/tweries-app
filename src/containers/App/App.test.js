import {
  cleanup,
  render,
  fireEvent,
  waitForDomChange
} from '@testing-library/react';
import FeatureProvider, { setFeatures } from 'feature-provider';
import React from 'react';
import Footer from '../../components/Footer/Footer';
import NavBar from '../../components/NavBar/NavBar';
import makeReducer from '../../store/makeReducer';
import features from '../../features';
import { useAuth0 } from '../../react-auth0-wrapper';
import App from './App';
import makeInitialState from '../../store/makeInitialState';
import { HIDE_TAGS_V1 } from '../../constants';

jest.mock('../../components/Footer/Footer');
jest.mock('../../components/NavBar/NavBar');
jest.mock('../../react-auth0-wrapper');

const feature = setFeatures(features);

const user = {
  name: 'tweries-app',
  picture:
    'https://abs.twimg.com/sticky/default_profile_images/default_profile_normal.png',
  sub: 'twitter|1183836409850814464'
};

describe('App', () => {
  beforeEach(() => {
    fetch.mockResponseOnce(JSON.stringify({ data: { message: 'baz' } }));
    Footer.mockImplementation(() => <div>Footer</div>);
    NavBar.mockImplementation(() => <div>NavBar</div>);
  });

  afterEach(() => {
    fetch.resetMocks();
    cleanup();
  });

  it('loading', () => {
    useAuth0.mockImplementation(() => ({
      loading: true
    }));

    const { container } = render(
      <FeatureProvider features={features}>
        <App
          initialState={makeInitialState({ feature })}
          reducer={makeReducer(feature)}
        />
      </FeatureProvider>
    );

    expect(container).toMatchSnapshot();
  });

  it('unauthenticated user', () => {
    useAuth0.mockImplementation(() => ({
      isAuthenticated: false
    }));

    const { container } = render(
      <FeatureProvider features={features}>
        <App
          initialState={makeInitialState({ feature })}
          reducer={makeReducer(feature)}
        />
      </FeatureProvider>
    );

    expect(container).toMatchSnapshot();
  });

  it('log in', () => {
    const mockLoginWithRedirect = jest.fn();
    useAuth0.mockImplementation(() => ({
      loginWithRedirect: mockLoginWithRedirect
    }));

    const { getByTestId } = render(
      <FeatureProvider features={features}>
        <App
          initialState={makeInitialState({ feature })}
          reducer={makeReducer(feature)}
        />
      </FeatureProvider>
    );

    fireEvent.click(getByTestId('login'));

    expect(mockLoginWithRedirect).toBeCalled();
  });

  it('generate tweetstorm and dismiss notification', async () => {
    fetch.mockResponseOnce(JSON.stringify({ data: { message: 'qux' } })); // INFO: adding a second response
    useAuth0.mockImplementation(() => ({
      isAuthenticated: true,
      user
    }));

    const { container, getByTestId } = render(
      <FeatureProvider
        features={features.filter(value => value !== HIDE_TAGS_V1)}
      >
        <App
          initialState={makeInitialState({ feature })}
          reducer={makeReducer(feature)}
        />
      </FeatureProvider>
    );
    fireEvent.change(getByTestId('source'), { target: { value: 'foo' } });
    fireEvent.change(getByTestId('hashtags'), { target: { value: '#bar' } });
    await waitForDomChange();
    fireEvent.click(getByTestId('tweet'));
    await waitForDomChange();
    fireEvent.click(getByTestId('dismiss'));

    expect(container).toMatchSnapshot();
  });
});

describe('App - errors', () => {
  it('fetchHealth error', () => {
    fetch.resetMocks();
    fetch.mockRejectOnce(new Error('Oh Noes!'));
    Footer.mockImplementation(() => <div>Footer</div>);
    NavBar.mockImplementation(() => <div>NavBar</div>);

    useAuth0.mockImplementation(() => ({
      isAuthenticated: false
    }));
    const mockReducer = jest.fn((state, action) =>
      makeReducer(feature)(state, action)
    );

    render(
      <FeatureProvider features={features}>
        <App
          initialState={makeInitialState({ feature })}
          reducer={mockReducer}
        />
      </FeatureProvider>
    );

    // TODO: add assertion
  });

  it('fetchTweetstorm error', async () => {
    fetch.resetMocks();
    fetch.mockResponseOnce(JSON.stringify({ data: { message: 'baz' } }));
    fetch.mockRejectOnce(new Error('Oh Noes!'));
    Footer.mockImplementation(() => <div>Footer</div>);
    NavBar.mockImplementation(() => <div>NavBar</div>);

    useAuth0.mockImplementation(() => ({
      isAuthenticated: true,
      user
    }));
    const mockReducer = jest.fn((state, action) =>
      makeReducer(feature)(state, action)
    );

    const { getByTestId } = render(
      <FeatureProvider features={features}>
        <App
          initialState={makeInitialState({ feature })}
          reducer={mockReducer}
        />
      </FeatureProvider>
    );
    fireEvent.change(getByTestId('source'), { target: { value: 'foo' } });
    await waitForDomChange();
    fireEvent.click(getByTestId('tweet'));
    await waitForDomChange();

    // TODO: add assertion
  });
});
