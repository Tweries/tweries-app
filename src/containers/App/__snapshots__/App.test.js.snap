// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App to match snapshot 1`] = `
<div>
  <article>
    <h1>
      Serial Twitter
    </h1>
    <form>
      <textarea
        data-testid="source"
        rows="16"
      >
        Mama, just killed a man
Put a gun against his head
Pulled my trigger, now he's dead
Mama, life had just begun
But now I've gone and thrown it all away
[..]
Mama, ooh
Didn't mean to make you cry
If I'm not back again this time tomorrow
Carry on, carry on
As if nothing really matters
[..]
Too late, my time has come
Sends shivers down my spine
Body's aching all the time
Goodbye everybody, I've got to go
Gotta leave you all behind and face the truth
      </textarea>
      <input
        data-testid="hashtags"
        type="text"
        value="#QueenFTW"
      />
      <button
        data-testid="generate"
      >
        Generate Tweetstorm
      </button>
      <ul />
      <button
        data-testid="tweet"
      >
        Tweet
      </button>
    </form>
    <footer>
      v
      0.3.0
    </footer>
  </article>
</div>
`;
